Single pe extractor

relleums opened the pull request:
Extract single photon pulses on the pixel time lines. 
- Calculates a list of arrival time slices of photon pulses on a time line.
- Has Unit tests.

This is a port of  a python implementation were I developed the algorithm and performed some toy simulation studies.

mackaiver commented on src/main/java/fact/features/singlePulse/timeLineExtraction/AddFirstArrayToSecondArray.java:
```diff
@@ -0,0 +1,21 @@
+        // endpoint of injection in f2: e2
+        int e2 = at + f1.length;
+        // naive endpoint of sampling in f1
+        int e1 = f1.length;
```
I don't think this variable is ever used. Can it be removed?

relleums replied:
This turned out to be a bug. I tracked it down with two more test cases.
Looks like I missed sth. from the python port.

mackaiver commented:
Looks nice. I removed some unused imports and stuff. Do you want me to push these changes here? Otherwise I create another pull request.

relleums commented:
Good catch sir. I think this is because I do not use any IDE.
I will work over the branch once more myself and try to make the comment blocks more like you make them look.

MaxNoe commented on src/main/java/fact/features/singlePulse/timeLineExtraction/AddFirstArrayToSecondArray.java:
```diff
@@ -0,0 +1,57 @@
+
+        if(end >= second.length) end = second.length;
+            // end is limited to length of second
+            // first array is completley before second. No overlap.
```
copy paste error?

MaxNoe commented:
I just threw it on the test file, there is a nice correlation between the number of found photons and the photoncharge. However, this method seems to tend to find more photons than the photoncharge..

![photoncounts](https://cloud.githubusercontent.com/assets/5488440/17210628/eb640c74-54bb-11e6-9803-74f63ac28330.png)

relleums commented:
The correlation looks actually better than I expected it to be, given that the 'fine tuning' of the single pe extractor was only done in my python MC toy simulation so far. Anyhow, I did not longer want to keep the single pe extractor to myself. We can now together investigate the fine tuning on observed data.
We need a tool in fact-tools where we can simulate time lines with pulses on them. This way we can compare all our extractors VS. the MC truth. So photoncharge vs MC truth and photonArrivalsCounts vs MC truth.

MaxNoe commented:
Yes. Really nice work.

I would opt for merging.

Could you maybe also provide a small discription in the CHANGELOG.md and increment the version number?

dneise commented:
Yes, the single pe extractor finds also singles outside the integration window. So the sum of found singles must always be larger or equal than the photon charge.

relleums commented:
I just saw that I forgot to copy an important step of my python sandbox (normalizing the convolution kernel). Now the performance for small pulses (1-5) is more accurate. I also simplified the interface and added unit tests for the overall SinglePulseExtractor.

mackaiver commented:
Ready to merge right?

MaxNoe commented:
LGTM

MaxNoe commented on examples/viewer.xml:
```diff
@@ -20,6 +20,12 @@
+        dataKey="DataCalibrated"
+        outputKey="PhotonArrivals"
+        maxIterations="50"
+    />
```
I would take this out here. Viewer should kind of be the minimal example.

MaxNoe commented on src/main/java/fact/extraction/SinglePulseExtraction.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public String getDataKey() {
```
We do not need getters. They just ad clutter.

jebuss commented on src/main/java/fact/extraction/SinglePulseExtraction.java:
```diff
@@ -0,0 +1,152 @@
+
+        npix = (Integer) input.get("NPIX");
+        roi = (Integer) input.get("NROI");
+        double[] timeLines = (double[]) input.get(dataKey);
```
To be consistent with other processors I suggest to call this array "data"

jebuss commented on src/main/java/fact/features/singlePulse/timeLineExtraction/ArgMax.java:
```diff
@@ -0,0 +1,27 @@
+    */
+    public ArgMax(double[] arr) {
+        arg = 0;
+        max = arr[0];;        
```
double ";"

jebuss commented on src/main/java/fact/features/singlePulse/timeLineExtraction/ArgMax.java:
```diff
@@ -0,0 +1,27 @@
+* Finds the maximum on an array and stores its position (arg) 
+* and its value (max).
+*/
+public class ArgMax {
```
shouldnt it be ArrMax or ArrayMax?

jebuss commented on src/main/java/fact/features/singlePulse/timeLineExtraction/TemplatePulse.java:
```diff
@@ -0,0 +1,66 @@
+        double[] template = new double[lengthInSlices];
+        for (int i = 0; i < time.length; i++) {
+
+            final double amplitude = 
```
Maybe add a comment with the source of this parameterization?

jebuss commented:
LGTM2

MaxNoe commented:
The `ArgMax` class does not return anything. Maybe a static method returning something would be better.
